### ISSUE TITLE
Implement support for Firefox OS

### ISSUE THREAD

**Issue by zalun** [NativeScript/nativescript-cli#517]
I've started to dig inside native-cli. Mostly a dummy approach to add platform firefoxos. It throws an error where it is looking for manifest.webapp, which is expected. Please let me know if I'm on the right path.

https://github.com/NativeScript/nativescript-cli/compare/master...zalun:firefoxos-support

**Comment by fealebenpae** [comment]
Hi @zalun, this looks good so far.

**Comment by valentinstoychev** [comment]
@zalun - great, please keep us posted on the progress. 

What is the plan for the UI creating? Do you plan to translate the {N} UI declaration to Firefox UI?

**Comment by zalun** [comment]
Whatever the default way is. I want to minimize the differences between FxOS and other platforms. Fortunately all of our UI is opensource, it's not complicated to copy that.

**Comment by zalun** [comment]
At the moment I struggle with the system itself. How the code flows etc. I don't want to implement any work arounds and most of the time I need to play with backward engineering. It would be great to avoid that.

**Comment by zalun** [comment]
add platform firefoxos is working (at least it doesn't throw)
build is not doing a thing (except of printng an info that it's happening) as there is nothing more in the buildProject for now.
The idea is to compile to JavaScript and zip the directory with manifest file.
I've created a http://github.com/zalun/firefox-runtime which is currently a framework directory with the manifest file.

**Comment by zalun** [comment]
As of the UI translation - there is nothing like FirefoxOS UI, everything except of dialog boxes needs to be (re)created in {N} FirefoxOS support

**Comment by zalun** [comment]
@fealebenpae @valentinstoychev - I'm looking for bit's about exporting the XML to HTML - has this been in work already? Should I start to play with plugins already?

**Comment by zalun** [comment]
This is what I've found out from code so far

add platform
-----------------
* firefox-runtime project
  * contents of the  ``framework`` directory is copied to the  ``platform//firefoxos/``
    
* lib/services/platform-service.ts :: addPlatform / addPlatformCore
* lib/services/firefox-project-service.ts :: interpolateData
  * manifest is updated with data stored in [app]/app/package.json
    
build
--------
* lib/services/platform-service:: buildPlatform
   * lib/services/platform-service:: preparePlatform
      * copy app folder to native project to platforms/firefoxos/assets
      * copy App_Resources directory - platforms/firefoxos/assets/app/App_Resources/FirefoxOS/ * (formerly copied with app directory) to platforms/firefoxos/res
      * remove platforms/firefoxos/assets/app/App_Resources
   * lib/services/platform-service:: processPlatformSpecificFiles
      * Renames the files that have `platform` as substring and removes the files from other platform
   * lib/services/firefoxos-project-service.ts :: buildProject
     * TODO: compile HTML
     * TODO: zip  to package

**Comment by Nathanaela** [comment]
@zalun -- I'm not with Telerik; but I've dug into a lot of things in the platform.     Your summary above looks correct on the command line tool from what I've seen; but someone at Telerik would obviously know the best.

What I did want to comment on is that I think you are going down the wrong trail with the XML files.  You don't need to convert them to anything or to html.      The XML files are just a layout; they are NOT used for anything but generating the JavaScript that makes the screen -- and that code is already written for you.   If you look at the "tns_modules/ui/builder" and "tns_modules/xml" you will see that the builder all it does is read the xml file; parse it and then "require" any JavaScript modules that the xml references.   So if it see:
<MyComponent blah="1" hi="2"> it will require ("MyComponent") and then create a new MyComponent and then set MyComponent.blah = "1" and MyCompnent.hi = "2" since that is what you put in the .xml file.

   Basically what you need to do for the display components is go through the tns_modules/ui folder and create a javascript code that will create the equivlent UI element in the FXos.   Once you have that; most of the UI platform should work pretty well for you.   

You might also dig a bit into the View code module which is in ui/core as that is the base from pretty much all ui.

**Comment by listepo** [comment]
+1

**Comment by ligaz** [comment]
As of December last year (2015) Firefox OS is considered [dead](https://discourse.mozilla-community.org/t/firefox-os-connected-devices-announcement/6864). :cry:
